V4 docs (#225)
moodlehq/moodle-plugin-ci: docs/UPGRADE-4.0.md

---
layout: page
title: Upgrading from 3.x to 4.0
---

This document outlines the steps one should take when upgrading to the new major version.

## Step 1: Review the new requirements

Requirements have changed, this project now requires **PHP 7.4 or later** and **Moodle 3.8.3** or later.

## Step 2: Review the change log

Detailed information about what changed in Version 4 can be found in the [change log](CHANGELOG.md).

## Step 3: Review the CIs configuration files

### GitHub Actions

Review the updated [gha.dist.yml](https://github.com/moodlehq/moodle-plugin-ci/blob/master/gha.dist.yml)
and update the `.github/workflows/ci.yml` (or whatever it's named) file in your plugin. For detailed information about the contents of that `ci.yml` file, please see this [help document](GHAFileExplained.md).

> A summary of the actions required, really simple for version 4, is:

Reviewer: I like the idea of celebrating the simplicity, but I also shy away from calling something "simple" if I'm giving someone else instructions (in case it's not simple for them).
```suggestion
A summary of the actions required for version 4 is:
```

> 1. Change the `composer create-project` line and bump to from version 3 (^3) to version 4 (^4).

Reviewer: ```suggestion
1. Change the `composer create-project` line and bump from version 3 (`^3`) to version 4 (`^4`).
```

2. Test it and, hopefully, party!

### Travis CI

Review the updated [.travis.dist.yml](https://github.com/moodlehq/moodle-plugin-ci/blob/master/.travis.dist.yml)
and update your `.travis.yml` file in your plugin. For detailed information about the contents of `.travis.dist.yml`

> file, please see this [help document](TravisFileExplained.md).

Reviewer: Using self-describing strings for links and avoiding using the same string for two different links are important accessibility criteria. 
```suggestion
file, please see our [sample Travis workflow file explanation](TravisFileExplained.md).
```

> 1. Change the `composer create-project` line and bump to from version 3 (^3) to version 4 (^4).

Reviewer: Makes it a bit more clear which string literal should be used.
```suggestion
1. Change the `composer create-project` line and bump from version 3 (`^3`) to version 4 (`^4`).
```

2. Test it and, hopefully, party!

## FAQ

### What is happening to Version 3?

Version 3 still exists within its own [3.x](https://github.com/moodlehq/moodle-plugin-ci/tree/3.x) branch. You can continue using it, but it is no longer getting new features and may not receive additional updates.  In addition, it may start breaking in Moodle 4.3 or later, because of 8.2 incompatibilities.

### Why Version 4?

> Mainly to be able to modernise all the internal components of the application that, basically were the same since version 2, making it really future-ready. And to leave behind very old versions of PHP (< 7.4) and Moodle (< 3.8.3), not supported since long ago.

Reviewer: ```suggestion
Mainly to be able to modernise all the internal components of the application which had remained largely unmodified since version 2. This will make it much easier to plan for the future and to leave behind very old versions of PHP (< 7.4) and Moodle (< 3.8.3), which have not been supported for a long time.
```

### Can I run Version 3 and 4?

> If your plugin uses a single Git branch, but still works on Moodle versions older than 3.8.3, then you might be wanting to try to use both versions in a single GiHub actions or Travis file.

Reviewer: ```suggestion
If your plugin uses a single Git branch, but still works on Moodle versions older than 3.8.3, then you might want to try using both versions in a single GiHub Action or Travis file.
```

First of all, this is not recommended due to its complexity. The preferred method is to use separate branches so your CI files are only using one version or another.

> Despite all the above if, still, you want to keep both versions working together, you can, via environment variables, decide which version is used for every job defined.

Reviewer: ```suggestion
Despite our recommendations, if you still want to run both versions in the same CI process, you can set environment variables to specify which version to use for every job.
```
